libcasr/src/java.rs

```diff
@@ -0,0 +1,272 @@
+            blocks.push(JavaExceptionBlock { body, footer });
+        }
+        if blocks.is_empty() {
+            return Err(Error::Casr("Couldn't find stacktrace".to_string()));
+        }
+        let last_block = blocks.last().unwrap();
+        let Some(mut prev_num) = last_block.footer else {
+            let stacktrace = last_block.body.iter().rev().map(|x| x.to_string()).collect::<Vec<String>>();
+            if stacktrace.is_empty() {
+                return Err(Error::Casr(
+                    "Empty stacktrace.".to_string()
+                ));
+            }
+            *JAVA_CRASH_LINE.lock().unwrap() = stacktrace[0].clone();
+            return Ok(stacktrace);
+        };
+        let mut backward_stacktrace = Vec::new();
+        let mut forward_stacktrace: Vec<&str> = Vec::new();
+        forward_stacktrace
+            .extend_from_slice(&last_block.body.iter().rev().copied().collect::<Vec<&str>>());
+
+        for block in blocks.iter().rev().skip(1) {
+            if let Some(cur_num) = block.footer {
+                let diff = prev_num - cur_num;
+                forward_stacktrace.extend_from_slice(
+                    &block.body[..diff]
+                        .iter()
+                        .rev()
+                        .copied()
+                        .collect::<Vec<&str>>(),
+                );
+                backward_stacktrace.extend_from_slice(&block.body[diff..]);
+                prev_num = cur_num;
+            } else {
+                forward_stacktrace.extend_from_slice(
+                    &block.body[..prev_num]
+                        .iter()
+                        .rev()
+                        .copied()
+                        .collect::<Vec<&str>>(),
+                );
+                backward_stacktrace.extend_from_slice(&block.body[prev_num..]);
+                break;
+            }
+        }
+
+        backward_stacktrace.reverse();
+        backward_stacktrace.extend_from_slice(forward_stacktrace.as_slice());
+        if forward_stacktrace.is_empty() {
+            return Err(Error::Casr("Empty stacktrace.".to_string()));
+        }
+        *JAVA_CRASH_LINE.lock().unwrap() = forward_stacktrace[0].to_string();
+
+        Ok(backward_stacktrace.iter().map(|x| x.to_string()).collect())
+    }
+
+    fn parse_stacktrace(entries: &[String]) -> Result<Stacktrace> {
+        let mut stacktrace = Stacktrace::new();
+        let re = Regex::new(r"(?:\s|\t)*at (.*)\((.*)\)").unwrap();
+        let rd = Regex::new(r"(.*):(\d+)").unwrap();
+
+        for entry in entries.iter() {
+            let Some(cap) = re.captures(entry) else {
+                return Err(Error::Casr(format!(
+                    "Couldn't parse stacktrace line: {entry}")
+                ));
+            };
+            let debug = cap.get(2).unwrap().as_str().to_string();
+            if debug.contains("Unknown Source") || debug.contains("Native Method") {
```

Comment:
Aren't native methods from C++?

Reply:
yes, they are. They should be added in `java_ignored_frames...` in the future.